File: ai-backend/models/ticket.py
"""
Data models for ticket assignment workflow
Updated to match the actual database schema
"""
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field, validator
from datetime import datetime
from enum import Enum
from .skill import Skill

class TicketStatus(str, Enum):
    """Ticket status enumeration"""
    NEW = "new"
    ASSIGNED = "assigned"
    IN_PROGRESS = "in_progress"
    ON_HOLD = "on_hold"
    RESOLVED = "resolved"
    CLOSED = "closed"
    CANCELLED = "cancelled"

class PriorityLevel(str, Enum):
    """Priority level enumeration"""
    LOW = "low"
    NORMAL = "normal"
    HIGH = "high"
    CRITICAL = "critical"

class ImpactLevel(str, Enum):
    """Impact level enumeration"""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"

class UrgencyLevel(str, Enum):
    """Urgency level enumeration"""
    LOW = "low"
    NORMAL = "normal"
    HIGH = "high"
    CRITICAL = "critical"

class ComplexityLevel(str, Enum):
    """Complexity level enumeration"""
    LEVEL_1 = "level_1"
    LEVEL_2 = "level_2"
    LEVEL_3 = "level_3"

class Ticket(BaseModel):
    """Ticket model matching the database schema"""
    # Primary fields
    id: Optional[int] = Field(None, description="Ticket unique identifier")
    subject: str = Field(..., min_length=5, max_length=500, description="Ticket subject/title")
    description: str = Field(..., description="Ticket description")
    
    # Status and categorization
    status: TicketStatus = Field(default=TicketStatus.NEW, description="Current ticket status")
    priority: PriorityLevel = Field(default=PriorityLevel.NORMAL, description="Ticket priority")
    impact: ImpactLevel = Field(default=ImpactLevel.MEDIUM, description="Business impact level")
    urgency: UrgencyLevel = Field(default=UrgencyLevel.NORMAL, description="Urgency level")
    complexity_level: ComplexityLevel = Field(default=ComplexityLevel.LEVEL_1, description="Technical complexity")
    
    # Tags and categorization
    tags: Optional[List[str]] = Field(default=[], description="Ticket tags")
    
    # SLA and timing fields
    sla_violated: bool = Field(default=False, description="Whether SLA has been violated")
    resolution_due: Optional[datetime] = Field(None, description="When the ticket should be resolved")
    resolution_date: Optional[datetime] = Field(None, description="When the ticket was actually resolved")
    first_response_time: Optional[int] = Field(None, ge=0, description="First response time in minutes")
    response_time: Optional[int] = Field(None, ge=0, description="Response time in minutes")
    resolution_time: Optional[int] = Field(None, ge=0, description="Resolution time in minutes")
    
    # Escalation and tracking
    escalation_count: int = Field(default=0, ge=0, description="Number of times ticket was escalated")
    reopened_count: int = Field(default=0, ge=0, description="Number of times ticket was reopened")
    
    # Relationships
    requester_id: int = Field(..., description="ID of the user who created the ticket")
    assigned_technician_id: Optional[int] = Field(None, description="ID of the assigned technician")
    
    # JSON fields for complex data
    required_skills: Optional[List[Dict[str, Any]]] = Field(default=[], description="Skills required for this ticket")
    tasks: Optional[List[Dict[str, Any]]] = Field(default=[], description="Tasks associated with this ticket")
    work_logs: Optional[List[Dict[str, Any]]] = Field(default=[], description="Work logs for this ticket")
    audit_trail: Optional[List[Dict[str, Any]]] = Field(default=[], description="Audit trail entries")
    attachments: Optional[List[Dict[str, Any]]] = Field(default=[], description="File attachments")
    
    # Timestamps
    first_response_at: Optional[datetime] = Field(None, description="When first response was given")
    resolved_at: Optional[datetime] = Field(None, description="When ticket was resolved")
    closed_at: Optional[datetime] = Field(None, description="When ticket was closed")
    created_at: Optional[datetime] = Field(default_factory=datetime.now, description="When ticket was created")
    updated_at: Optional[datetime] = Field(default_factory=datetime.now, description="When ticket was last updated")
    
    # Customer satisfaction
    satisfaction_rating: Optional[int] = Field(None, ge=1, le=5, description="Customer satisfaction rating (1-5)")
    feedback: Optional[str] = Field(None, description="Customer feedback")
    
    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }
        use_enum_values = True

class SkillScore(BaseModel):
    """Skill with relevance score - updated to use proper Skill model"""
    skill: Skill = Field(..., description="The skill object")
    score: float = Field(..., ge=0.0, le=1.0, description="Relevance score from 0.0 to 1.0")
    
    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }

class SkillScoreSimple(BaseModel):
    """Simplified skill score for API responses"""
    skill_name: str = Field(..., description="The skill name")
    skill_id: Optional[int] = Field(None, description="The skill ID")
    score: float = Field(..., ge=0.0, le=1.0, description="Relevance score from 0.0 to 1.0")

class Technician(BaseModel):
    """Technician model"""
    id: int = Field(..., description="Technician unique identifier")
    name: str = Field(..., description="Technician name")
    skills: List[str] = Field(..., description="List of technician skills")
    availability: str = Field(..., description="Current availability status")
    rating: float = Field(..., ge=0.0, le=5.0, description="Technician rating from 0.0 to 5.0")
    experience_years: Optional[int] = Field(None, description="Years of experience")
    complexity_level: Optional[ComplexityLevel] = Field(None, description="Technician's complexity level")
    current_workload: Optional[int] = Field(None, ge=0, description="Current number of active tickets")

class TicketAssignmentRequest(BaseModel):
    """Request model for ticket assignment"""
    ticket: Ticket

class TicketAssignmentResponse(BaseModel):
    """Response model for ticket assignment"""
    success: bool = Field(..., description="Whether the assignment was successful")
    selected_technician_id: Optional[int] = Field(None, description="Selected technician ID")
    justification: Optional[str] = Field(None, description="Justification for the selection")
    error_message: Optional[str] = Field(None, description="Error message if assignment failed")
    
class TicketSummary(BaseModel):
    """Simplified ticket model for API requests"""
    subject: str = Field(..., min_length=5, max_length=500, description="Ticket subject")
    description: str = Field(..., description="Ticket description")
    priority: PriorityLevel = Field(default=PriorityLevel.NORMAL, description="Ticket priority")
    impact: ImpactLevel = Field(default=ImpactLevel.MEDIUM, description="Business impact")
    urgency: UrgencyLevel = Field(default=UrgencyLevel.NORMAL, description="Urgency level")
    complexity_level: ComplexityLevel = Field(default=ComplexityLevel.LEVEL_1, description="Technical complexity")
    tags: Optional[List[str]] = Field(default=[], description="Ticket tags")
    requester_id: int = Field(..., description="ID of the requester") 

File: backend/models/Ticket.js
'use strict';

module.exports = (sequelize, DataTypes) => {
  const Ticket = sequelize.define('Ticket', {
    id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true,
      allowNull: false
    },
    subject: {
      type: DataTypes.STRING(500),
      allowNull: false,
      validate: {
        notEmpty: true,
        len: [5, 500]
      }
    },
    description: {
      type: DataTypes.TEXT,
      allowNull: false,
      validate: {
        notEmpty: true
      }
    },
    status: {
      type: DataTypes.ENUM('new', 'assigned', 'in_progress', 'on_hold', 'resolved', 'closed', 'cancelled'),
      allowNull: false,
      defaultValue: 'new'
    },
    tags: {
      type: DataTypes.JSON,
      allowNull: true,
      defaultValue: []
    },
    priority: {
      type: DataTypes.ENUM('low', 'normal', 'high', 'critical'),
      allowNull: false,
      defaultValue: 'normal'
    },
    impact: {
      type: DataTypes.ENUM('low', 'medium', 'high', 'critical'),
      allowNull: false,
      defaultValue: 'medium'
    },
    urgency: {
      type: DataTypes.ENUM('low', 'normal', 'high', 'critical'),
      allowNull: false,
      defaultValue: 'normal'
    },
    // SLA related fields
    sla_violated: {
      type: DataTypes.BOOLEAN,
      allowNull: false,
      defaultValue: false
    },
    resolution_due: {
      type: DataTypes.DATE,
      allowNull: true
    },
    resolution_date: {
      type: DataTypes.DATE,
      allowNull: true
    },
    first_response_time: {
      type: DataTypes.INTEGER, // in minutes
      allowNull: true
    },
    response_time: {
      type: DataTypes.INTEGER, // in minutes
      allowNull: true
    },
    resolution_time: {
      type: DataTypes.INTEGER, // in minutes
      allowNull: true
    },
    escalation_count: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 0
    },
    // Relationships
    requester_id: {
      type: DataTypes.INTEGER,
      allowNull: false,
      references: {
        model: 'users',
        key: 'id'
      }
    },
    assigned_technician_id: {
      type: DataTypes.INTEGER,
      allowNull: true,
      references: {
        model: 'technicians',
        key: 'id'
      }
    },
    // JSON fields for simple data storage
    required_skills: {
      type: DataTypes.JSON,
      allowNull: true,
      defaultValue: [],
      comment: 'Array of skill IDs: [1, 2, 3]'
    },
    tasks: {
      type: DataTypes.JSON,
      allowNull: true,
      defaultValue: [],
      comment: 'Array of task objects: [{sub: "Task 1", status: "pending", description: "Details"}]'
    },
    work_logs: {
      type: DataTypes.JSON,
      allowNull: true,
      defaultValue: [],
      comment: 'Array of work log entries with timestamps and notes'
    },
    audit_trail: {
      type: DataTypes.JSON,
      allowNull: true,
      defaultValue: [],
      comment: 'Array of all changes made to the ticket'
    },
    // Additional tracking
    first_response_at: {
      type: DataTypes.DATE,
      allowNull: true
    },
    resolved_at: {
      type: DataTypes.DATE,
      allowNull: true
    },
    closed_at: {
      type: DataTypes.DATE,
      allowNull: true
    },
    reopened_count: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 0
    },
    satisfaction_rating: {
      type: DataTypes.INTEGER,
      allowNull: true,
      validate: {
        min: 1,
        max: 5
      }
    },
    score: {
      type: DataTypes.FLOAT,
      allowNull: true,
      validate: {
        min: 0.0,
        max: 10.0
      }
    },
    justification: {
      type: DataTypes.TEXT,
      allowNull: true,
      defaultValue: 'Severity: Medium',
      comment: 'Justification for ticket decisions or actions'
    },
    feedback: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    created_at: {
      type: DataTypes.DATE,
      allowNull: false,
      defaultValue: DataTypes.NOW
    },
    updated_at: {
      type: DataTypes.DATE,
      allowNull: false,
      defaultValue: DataTypes.NOW
    }
  }, {
    tableName: 'tickets',
    timestamps: true,
    underscored: true,
    indexes: [
      {
        fields: ['status']
      },
      {
        fields: ['priority']
      },
      {
        fields: ['urgency']
      },
      {
        fields: ['impact']
      },
      {
        fields: ['requester_id']
      },
      {
        fields: ['assigned_technician_id']
      },
      {
        fields: ['sla_violated']
      },
      {
        fields: ['created_at']
      },
      {
        fields: ['resolution_due']
      }
    ]
  });

  Ticket.associate = function(models) {
    // Simple relationships - Ticket belongs to a requester (User)
    Ticket.belongsTo(models.User, {
      foreignKey: 'requester_id',
      as: 'requester'
    });
    
    // Simple relationship - Ticket can be assigned to a technician
    Ticket.belongsTo(models.Technician, {
      foreignKey: 'assigned_technician_id',
      as: 'assigned_technician'
    });
  };

  return Ticket;
};